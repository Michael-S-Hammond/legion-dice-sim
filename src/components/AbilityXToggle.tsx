import React from 'react';

type AbilityXProps = {
    id: string,
    label: string,
    visible: boolean,
    active: boolean,
    onActiveChanged: Function,
    value: number,
    onValueChanged: Function,
    maxValue: number,
}

class AbilityXToggle extends React.Component<AbilityXProps> {
    constructor(props : AbilityXProps) {
        super(props);
    }
    
    handleValueChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
        const newValue = Number(e.target.value);
        this.props.onValueChanged(newValue);
    }

    handleActiveChanged = (e: React.ChangeEvent<HTMLInputElement>) => {
        const newActive = e.target.checked;
        this.props.onActiveChanged(newActive);
    }

    render() {
        const options = [];
        for(let i = 1; i <= this.props.maxValue; i++) {
            options.push(<option key={`${this.props.id}-abilityxtoggle-option-${i}`} value={i}>{i}</option>)
        }
        return (
            <div key={`${this.props.id}-abilityxtoggle`} className={`${ this.props.visible ? 'd-flex collapse.show' : 'collapse'} justify-content-center my-3 custom-control custom-switch`}>
                <input key={`${this.props.id}-abilityxtoggle-input`} type="checkbox" className="custom-control-input my-auto"
                    id={this.props.id}
                    checked={this.props.active}
                    onChange={this.handleActiveChanged}></input>
                <label key={`${this.props.id}-abilityxtoggle-label`} className="custom-control-label drop-down-label mx-2 my-auto" htmlFor={this.props.id}>{this.props.label}</label>
                <select key={`${this.props.id}-abilityxtoggle-select`} value={this.props.value} className="rounded-lg mr-4 px-2"
                    onChange={this.handleValueChange}>
                    {options}
                </select>
            </div>
        );
    }
}

export default AbilityXToggle;
