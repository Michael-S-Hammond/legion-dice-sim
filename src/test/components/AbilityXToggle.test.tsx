import { shallow } from 'enzyme';
import { create } from 'react-test-renderer';

import AbilityXToggle from '../../components/AbilityXToggle';

describe('AbilityXToggle', () => {
    it('matches the snapshot', () => {
        function onActiveChanged() {}
        function onValueChanged() {}

        const counter = <AbilityXToggle
                id='testAbility'
                label='Test'
                visible={true}
                active={true}
                onActiveChanged={onActiveChanged}
                value={1}
                onValueChanged={onValueChanged}
                maxValue={4}
            ></AbilityXToggle>;
        const snapshot = create(counter);
        expect(snapshot.toJSON()).toMatchSnapshot();
    });

    it('matches the snapshot when hidden', () => {
        function onActiveChanged() {}
        function onValueChanged() {}

        const counter = <AbilityXToggle
                id='testAbility'
                label='Test'
                visible={false}
                active={true}
                onActiveChanged={onActiveChanged}
                value={1}
                onValueChanged={onValueChanged}
                maxValue={4}
            ></AbilityXToggle>;
        const snapshot = create(counter);
        expect(snapshot.toJSON()).toMatchSnapshot();
    });

    it('handles active changed', () => {
        let isActive = false;
        const onActiveChanged = jest.fn((newisActive) => {
            isActive = newisActive;
        });

        let value = 1;
        const onValueChanged = jest.fn((newValue) => {
            value = newValue;
        });

        const wrapper = shallow(<AbilityXToggle
                id='testAbility'
                label='Test'
                visible={true}
                active={false}
                onActiveChanged={onActiveChanged}
                value={3}
                onValueChanged={onValueChanged}
                maxValue={4}
            ></AbilityXToggle>);
        wrapper.find('.custom-control-input').simulate('change', { target: { checked: true }});

        expect(onActiveChanged).toHaveBeenCalledTimes(1);
        expect(isActive).toEqual(true);
        expect(onValueChanged).toHaveBeenCalledTimes(0);
    })

    it('handles value changed', () => {
        let isActive = false;
        const onActiveChanged = jest.fn((newisActive) => {
            isActive = newisActive;
        });

        let value = 1;
        const onValueChanged = jest.fn((newValue) => {
            value = newValue;
        });

        const wrapper = shallow(<AbilityXToggle
                id='testAbility'
                label='Test'
                visible={true}
                active={false}
                onActiveChanged={onActiveChanged}
                value={3}
                onValueChanged={onValueChanged}
                maxValue={5}
            ></AbilityXToggle>);
        wrapper.find('select').simulate('change', { target: { value: 2 }});

        expect(onActiveChanged).toHaveBeenCalledTimes(0);
        expect(onValueChanged).toHaveBeenCalledTimes(1);
        expect(value).toEqual(2);
    })
});
