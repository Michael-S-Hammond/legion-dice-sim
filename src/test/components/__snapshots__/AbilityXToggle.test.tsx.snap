// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbilityXToggle matches the snapshot 1`] = `
<div
  className="d-flex collapse.show justify-content-center my-3 custom-control custom-switch"
>
  <input
    checked={true}
    className="custom-control-input my-auto"
    id="testAbility"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="custom-control-label drop-down-label mx-2 my-auto"
    htmlFor="testAbility"
  >
    Test
  </label>
  <select
    className="rounded-lg mr-4 px-2"
    onChange={[Function]}
    value={1}
  >
    <option
      value={1}
    >
      1
    </option>
    <option
      value={2}
    >
      2
    </option>
    <option
      value={3}
    >
      3
    </option>
    <option
      value={4}
    >
      4
    </option>
  </select>
</div>
`;

exports[`AbilityXToggle matches the snapshot when hidden 1`] = `
<div
  className="collapse justify-content-center my-3 custom-control custom-switch"
>
  <input
    checked={true}
    className="custom-control-input my-auto"
    id="testAbility"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="custom-control-label drop-down-label mx-2 my-auto"
    htmlFor="testAbility"
  >
    Test
  </label>
  <select
    className="rounded-lg mr-4 px-2"
    onChange={[Function]}
    value={1}
  >
    <option
      value={1}
    >
      1
    </option>
    <option
      value={2}
    >
      2
    </option>
    <option
      value={3}
    >
      3
    </option>
    <option
      value={4}
    >
      4
    </option>
  </select>
</div>
`;
